sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -244,28 +244,21 @@ class NodeHttpClient implements HttpClient {
   }
 
   private getOrCreateAgent(request: PipelineRequest, isInsecure: boolean): http.Agent {
-    if (!request.disableKeepAlive) {
-      if (isInsecure) {
-        if (!this.httpKeepAliveAgent) {
-          this.httpKeepAliveAgent = new http.Agent({
-            keepAlive: true,
-          });
-        }
+    if (!request.tlsSettings && request.disableKeepAlive) {
+      return isInsecure ? http.globalAgent : https.globalAgent;
+    }
 
-        return this.httpKeepAliveAgent;
-      } else {
-        if (!this.httpsKeepAliveAgent) {
-          this.httpsKeepAliveAgent = new https.Agent({
-            keepAlive: true,
-          });
-        }
+    const agentOptions: https.AgentOptions = {
+      ...request.tlsSettings,
+      keepAlive: !request.disableKeepAlive,
+    };
 
-        return this.httpsKeepAliveAgent;
-      }
-    } else if (isInsecure) {
-      return http.globalAgent;
+    if (isInsecure) {
+      this.httpKeepAliveAgent = new http.Agent(agentOptions);
```

Comment:
we're never caching now that tls settings get passed in, which can cause some troubles (I believe pooling happens per-agent) 

Even though it's slightly more work, I think we should have something like

```ts
interface CachedAgent {
   tlsSettings: TlsSettings;
   agent: Agent;
}
```

that we can assign to `httpKeepAliveAgent` and `httpsKeepAliveAgent` -- then we just need a helper that checks if the new TLS settings are compatible with the previous ones. I think this would be sufficient for now, if we want to get fancy later and try to keep a set number of agents, we can revisit.

Reply:
I'm thinking we can set a uniqueId to the tlsSettings when creating the policy and use that to compare if the settings have changed and decide to cache a new agent or not. This is to avoid deep comparison every time

Reply:
If we tag the settings object we should do so with a WeakMap<> to avoid modifying the user created object. This also will help with removing the tag from the old settings object if a new one is passed in.